src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddleware.cs

```diff
@@ -113,11 +113,21 @@ private async Task HandleException(HttpContext context, ExceptionDispatchInfo ed
                 };
                 context.Features.Set<IExceptionHandlerFeature>(exceptionHandlerFeature);
                 context.Features.Set<IExceptionHandlerPathFeature>(exceptionHandlerFeature);
-                context.Response.StatusCode = 500;
+                context.Response.StatusCode = StatusCodes.Status500InternalServerError;
                 context.Response.OnStarting(_clearCacheHeadersDelegate, context.Response);
 
                 await _options.ExceptionHandler(context);
```

Comment:
It might be cleaner if the logic is inverted. Also, if the exception gets rethrown then we can rely on the server to do the natural cleanup and set the right status code.
```suggestion
                await _options.ExceptionHandler(context);
                
               if (context.Response.StatusCode != StatusCodes.Status404NotFound)
               {
                    if (_diagnosticListener.IsEnabled() && _diagnosticListener.IsEnabled("Microsoft.AspNetCore.Diagnostics.HandledException"))
                    {
                        _diagnosticListener.Write("Microsoft.AspNetCore.Diagnostics.HandledException", new { httpContext = context, exception = edi.SourceException });
                    }

                    return;
                }
                
                _logger.LogWarning("No exception handler was found, rethrowing original exception.");
```